File: 06/breadth-first_search_practice0601.py
from collections import deque

graph = {}
graph["S"] = ["A", "C"]
graph["A"] = ["B", "F"]
graph["B"] = []
graph["C"] = ["B", "D"]
graph["D"] = ["F"]

def search(start):
    search_queue = deque()
    search_queue += graph[start]
    searched = []
    while search_queue:
        position = search_queue.popleft()
        if not position in searched:
            if position == "F":
                return True
            else:
                search_queue += graph[position]
                searched.append(position)

    return False


print(search("S"))

File: 06/breadthFirstSearchPractice0601.js
const graph = {};

graph["S"] = ["A", "C"];
graph["A"] = ["B", "F"];
graph["B"] = [];
graph["C"] = ["B", "D"];
graph["D"] = ["F"];

const search = (start) => {
	let search_queue = [];
	search_queue = [...search_queue, ...graph[start]];

	let searched = [];

	while (search_queue.length !== 0) {
		const position = search_queue.shift();

		if (!searched.includes(position)) {
			if (position === "F") {
				return true;
			} else {
				search_queue = [...search_queue, ...graph[position]];
				searched = [...searched, position];
			}
		}
	}

	return false;
};

console.log(search("S"));
